safehttp/response_writer.go

```diff
@@ -51,52 +52,66 @@ func (w *ResponseWriter) Interceptor(key string) Interceptor {
 }
 
 // Result TODO
-type Result struct {
-	written bool
-}
+type Result struct{}
 
 // Write TODO
 func (w *ResponseWriter) Write(resp Response) Result {
+	w.markWritten()
 	if err := w.d.Write(w.rw, resp); err != nil {
 		panic("error")
 	}
-	return Result{written: true}
+	return Result{}
 }
 
 // WriteTemplate TODO
 func (w *ResponseWriter) WriteTemplate(t Template, data interface{}) Result {
+	w.markWritten()
 	if err := w.d.ExecuteTemplate(w.rw, t, data); err != nil {
 		panic("error")
 	}
-	return Result{written: true}
+	return Result{}
 }
 
 // ClientError TODO
 func (w *ResponseWriter) ClientError(code StatusCode) Result {
+	w.markWritten()
 	if code < 400 || code >= 500 {
 		// TODO(@mihalimara22): Replace this when we decide how to handle this case
 		panic("wrong method called")
 	}
 	http.Error(w.rw, http.StatusText(int(code)), int(code))
-	return Result{written: true}
+	return Result{}
 }
 
 // ServerError TODO
 func (w *ResponseWriter) ServerError(code StatusCode) Result {
+	w.markWritten()
 	if code < 500 || code >= 600 {
 		// TODO(@mattiasgrenfeldt, @mihalimara22, @kele, @empijei): Decide how it should
 		// be communicated to the user of the framework that they've called the wrong
 		// method.
-		return Result{written: true}
+		panic("wrong method called")
 	}
 	http.Error(w.rw, http.StatusText(int(code)), int(code))
-	return Result{written: true}
+	return Result{}
 }
 
 // Redirect responds with a redirect to a given url, using code as the status code.
 func (w *ResponseWriter) Redirect(r *IncomingRequest, url string, code StatusCode) Result {
+	w.markWritten()
+	if code < 300 || code >= 400 {
+		panic("wrong method called")
+	}
 	http.Redirect(w.rw, r.req, url, int(code))
-	return Result{written: true}
+	return Result{}
+}
+
+// markWritten sets written to true. If written was already true, it panics.
```

Comment:
Have a comment that describes how the function is useful not its exact implementation.